File: backend/src/main/java/es/udc/paproject/backend/model/services/PatternService.java
package es.udc.paproject.backend.model.services;

import es.udc.paproject.backend.model.entities.Pattern;
import es.udc.paproject.backend.model.entities.Section;
import es.udc.paproject.backend.model.entities.Tool;
import es.udc.paproject.backend.model.entities.Yarn;
import es.udc.paproject.backend.model.exceptions.*;

import java.math.BigDecimal;
import java.util.List;

public interface PatternService {


    Pattern createPattern(Long userId, Long craftId, Long subcategoryId, String title, String description,
                          BigDecimal price, Boolean active, String introduction, String notes, String gauge,
                          String sizing, int difficultyLevel, String time, String abbreviations,
                          String specialAbbreviations, String language,
                          List<String> imagesUrl, List<Tool> tools, List<Yarn> yarns, List<Section> sections)
            throws InstanceNotFoundException, UserNotSellerException, MaxItemsExceededException;


    /**
     * Returns all the patterns a particular user has added to their store
     */
    Block<Pattern> findAddedPatterns(Long userId, int page, int size) throws InstanceNotFoundException, UserNotSellerException;


    Pattern findPatternById(Long userId, Long productId) throws InstanceNotFoundException, UserNotOwnerException, PermissionException;




    Pattern editPattern(Long productId, Long userId, Long craftId, Long subcategoryId, String title, String description,
                        BigDecimal price, Boolean active, String introduction, String notes, String gauge,
                        String sizing, int difficultyLevel, String time, String abbreviations,
                        String specialAbbreviations, String language,
                        List<String> imagesUrl, List<Tool> tools, List<Yarn> yarns, List<Section> sections)
            throws InstanceNotFoundException, PermissionException, MaxItemsExceededException;

    void deletePattern(Long userId, Long productId) throws InstanceNotFoundException, PermissionException;

    Pattern findPurchasedPatternById(Long userId, Long patternId) throws InstanceNotFoundException, PermissionException;

    List<Pattern> findPurchasedPatterns(Long userId) throws InstanceNotFoundException;

}

File: frontend/src/backend/patternService.js
import {appFetch, config} from "./appFetch.js";

export const createPattern = (pattern, onSuccess, onErrors) =>
    appFetch(`/patterns`, config('POST', pattern), onSuccess, onErrors);

export const findAddedPatterns = ({page}, onSuccess) =>
    appFetch(`/patterns?page=${page}`, config('GET'), onSuccess);

export const findPatternById = (patternId, onSuccess) =>
    appFetch(`/patterns/${patternId}`, config('GET'), onSuccess);

export const editPattern = (pattern, onSuccess, onErrors) =>
    appFetch(`/patterns/${pattern.id}`, config('PUT', pattern),
        onSuccess, onErrors);

export const deletePattern = (patternId, onSuccess)=>
    appFetch(`/patterns/${patternId}`, config('DELETE'),
        onSuccess);


export const getPurchasedPatternById = (patternId, onSuccess) =>
    appFetch(`/patterns/purchased/${patternId}`, config('GET'), onSuccess);

export const getPurchasedPatterns = (onSuccess) =>
    appFetch(`/patterns/purchased`, config('GET'), onSuccess);